--- 

## Day 1: Trebuchet?! 

---

Something is wrong with global snow production, and you've been selected to take a look. The Elves have even given you a map; on it, they've used stars to mark the top fifty locations that are likely to be having problems.

You've been doing this long enough to know that to restore snow operations, you need to check all fifty stars by December 25th.

Collect stars by solving puzzles. Two puzzles will be made available on each day in the Advent calendar; the second puzzle is unlocked when you complete the first. Each puzzle grants one star. Good luck!

You try to ask why they can't just use a weather machine ("not powerful enough") and where they're even sending you ("the sky") and why your map looks mostly blank ("you sure ask a lot of questions") and hang on did you just say the sky ("of course, where do you think snow comes from") when you realize that the Elves are already loading you into a trebuchet ("please hold still, we need to strap you in").

As they're making the final adjustments, they discover that their calibration document (your puzzle input) has been amended by a very young Elf who was apparently just excited to show off her art skills. Consequently, the Elves are having trouble reading the values on the document.

The newly-improved calibration document consists of lines of text; each line originally contained a specific calibration value that the Elves now need to recover. On each line, the calibration value can be found by combining the first digit and the last digit (in that order) to form a single two-digit number.

For example:
```
1abc2
pqr3stu8vwx
a1b2c3d4e5f
treb7uchet
```

In this example, the calibration values of these four lines are `12`, `38`, `15`, and `77`. Adding these together produces 142.

Consider your entire calibration document. What is the sum of all of the calibration values?

To begin, get your puzzle input.

Answer: 
 

In [165]:
import pandas as pd

gold_stars = 0


In [166]:
calibration_document = pd.read_csv('./data/Day1.csv', header=0)

cd = calibration_document.copy()
cd.head()


elf_art
0                fivepqxlpninevh2xxsnsgg63pbvdnqptmg
1               eight8zlctbmsixhrvbpjb84nnmlcqkzrsix
2  hkxqfrqmsixfplbkpkdfzzszjxmdjtdkjlprrvr3gghlrq...
3                             zkjkctxvssix1dqb22five
4                                          4dtlmkfnm

In [182]:
def calibration_values(value):
  cal_nums = []
  [cal_nums.append(i) for i in value if i.isnumeric()]
  answer = int(cal_nums[0]+cal_nums[-1]) if len(cal_nums) !=0 else 0
  return answer
  # return int(cal_nums[0]+cal_nums[-1])

def sum_calibration_values(series):
  cal_vals = []
  [cal_vals.append(calibration_values(num)) for num in series]
  return sum(cal_vals)



In [183]:
test = cd['elf_art'][:5].copy()
sum_calibration_values(test)


196

In [184]:
test = ['1abc2', 'pqr3stu8vwx', 'a1b2c3d4e5f', 'treb7uchet']
sum_calibration_values(test)


142

In [185]:
sum_calibration_values(cd['elf_art'])


53651

Your puzzle answer was `53651`.

The first half of this puzzle is complete! It provides one gold star: *


In [171]:
def score_check():
  print("__you pull the map provided by the elves from your pocket to mark the current location as resolved.__")
  print("Upon determining the next destination to investigate, you note the current tally of gold stars collected is: ")
  return gold_stars + 1


In [172]:
score_check()


__you pull the map provided by the elves from your pocket to mark the current location as resolved.__
Upon determining the next destination to investigate, you note the current tally of gold stars collected is: 


1


---- 
### Part Two 
----

Your calculation isn't quite right. It looks like some of the digits are actually spelled out with letters: one, two, three, four, five, six, seven, eight, and nine also count as valid "digits".

Equipped with this new information, you now need to find the real first and last digit on each line. For example:
```
two1nine
eightwothree
abcone2threexyz
xtwone3four
4nineeightseven2
zoneight234
7pqrstsixteen
```

In this example, the calibration values are `29`, `83`, `13`, `24`, `42`, `14`, and `76`. Adding these together produces `281`.

What is the sum of all of the calibration values?

Answer:

In [173]:
num_dict = {
  'one': 1,
  'two': 2,
  'three': 3,
  'four': 4,
  'five': 5,
  'six': 6,
  'seven': 7,
  'eight': 8,
  'nine': 9
}


In [174]:
text_nums = list(num_dict.keys())
text_nums


['one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine']

In [175]:
def validate_text_nums(value):
  for i in list(num_dict.keys()):
    if i in value:
      value=value.replace(i, str(num_dict[i]))
    else:
      None
  return value


In [186]:
test = ['two1nine', 'eightwothree', 'abcone2threexyz', 'xtwone3four', '4nineeightseven2', 'zoneight234', '7pqrstsixteen']


In [189]:
for i in test:
  print(f'{i} ==> {calibration_values(i)}')
  print(f'{validate_text_nums(i)} ==> {calibration_values(validate_text_nums(i))}')
  print()


two1nine ==> 11
219 ==> 29

eightwothree ==> 0
eigh23 ==> 23

abcone2threexyz ==> 22
abc123xyz ==> 13

xtwone3four ==> 33
xtw134 ==> 14

4nineeightseven2 ==> 42
49872 ==> 42

zoneight234 ==> 24
z1ight234 ==> 14

7pqrstsixteen ==> 77
7pqrst6teen ==> 76



In [179]:
print('Expected value: 281')
sum_calibration_values(test)


Expected value: 281


211

In [177]:
validate_text_nums('4oneone14')


'41114'

In [178]:
test = [validate_text_nums(i) for i in test]


In [150]:
cd['cal_num'] = cd['elf_art'].apply(lambda x: validate_text_nums(x))
cd['num_art'] = cd['elf_art'].apply(lambda x: validate_text_nums(x))

cd.head()


elf_art  \
0                fivepqxlpninevh2xxsnsgg63pbvdnqptmg   
1               eight8zlctbmsixhrvbpjb84nnmlcqkzrsix   
2  hkxqfrqmsixfplbkpkdfzzszjxmdjtdkjlprrvr3gghlrq...   
3                             zkjkctxvssix1dqb22five   
4                                          4dtlmkfnm   

                                             num_art  
0                      5pqxlp9vh2xxsnsgg63pbvdnqptmg  
1                       88zlctbm6hrvbpjb84nnmlcqkzr6  
2  hkxqfrqm6fplbkpkdfzzszjxmdjtdkjlprrvr3gghlrqck...  
3                                  zkjkctxvs61dqb225  
4                                          4dtlmkfnm

In [151]:
sum_calibration_values(cd['num_art'])


53148

In [139]:
elf_art_nums_decoded = [validate_text_nums(i) for i in cd['elf_art']]


In [140]:
sum_calibration_values(elf_art_nums_decoded)


53148

In [133]:
s


'219'

In [98]:
s.replace(num_dict.keys(), num_dict.values())


TypeError: replace() argument 1 must be str, not dict_keys

In [ ]:
def calibration_values(value):
  cal_nums = []
  [cal_nums.append(i) for i in value if i.isnumeric()]
  return int(cal_nums[0]+cal_nums[-1])

def sum_calibration_values(series):
  cal_vals = []
  [cal_vals.append(calibration_values(num)) for num in series]
  return sum(cal_vals)

